### PR TITLE
[SMALLFIX] Do not write to temporary file for active journal log

### PR DESCRIPTION
**Pull request by madanadit**
Current journal entry log may not be closed before a crash. Make sure that the log is readable before a close by using CreateOptions.ensureAtomic = false.

### CODE REVIEW

**Comment by madanadit** [comment]
Jenkins, test this please

**Comment by madanadit** [comment]
@aaudiber @calvinjia Please take a look. Thanks!.

File: core/common/src/main/java/alluxio/underfs/options/CreateOptions.java [hunk 1]
```diff
@@ -24,24 +24,46 @@
...
    */
   public CreateOptions() {
+    mEnsureAtomic = true;
     mPermission = Permission.defaults().applyFileUMask();
   }
 
   /**
+   * @return ensure atomic
+   */
+  public boolean getEnsureAtomic() {
```
Comment (calvinjia) [inline core/common/src/main/java/alluxio/underfs/options/CreateOptions.java]: `isEnsureAtomic` for consistency?

**Comment by calvinjia** [comment]
@madanadit looks good overall, is it easy to add a test which verifies this behavior for the journal?

File: core/common/src/main/java/alluxio/underfs/options/CreateOptions.java [hunk 2]
```diff
@@ -61,17 +83,19 @@ public boolean equals(Object o) {
...
   @Override
   public int hashCode() {
-    return Objects.hashCode(mPermission);
+    return Objects.hashCode(mEnsureAtomic, mPermission);
   }
 
   @Override
   public String toString() {
     return Objects.toStringHelper(this)
+        .add("ensure_atomic", mEnsureAtomic)
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/options/CreateOptions.java]: ensure_atomic -> ensureAtomic

File: core/server/src/test/java/alluxio/master/journal/JournalWriterTest.java [hunk 3]
```diff
@@ -63,6 +65,8 @@ public void rotateLogOnFlushIOException() throws Exception {
     OutputStream mockOutStream = mock(OutputStream.class);
     UnderFileSystem mockUfs = mock(UnderFileSystem.class);
     doReturn(mockOutStream).when(mockUfs).create(mJournal.getCurrentLogFilePath());
```
Comment (aaudiber) [inline core/server/src/test/java/alluxio/master/journal/JournalWriterTest.java]: we can remove line 67 now

Comment (aaudiber) [inline core/server/src/test/java/alluxio/master/journal/JournalWriterTest.java]: same for the other two tests

**Comment by aaudiber** [comment]
I'm a bit unclear on the exact semantics for the stream returned by `UnderFileSystem#create`. Is calling `flush` on this stream meaningless by default? Why is atomicity the default behavior?

**Comment by madanadit** [comment]
@calvinjia @aaudiber Apparently all Journal integrations tests where broken as the LocalAlluxioCluster was started with a default WriteType.CACHE_THROUGH. Instead of recovering from log, Alluxio master was fetching created files from UFS during list (loadMetadata).

I've added a new test case to make sure that jounral is read correctly even when no close is called on EvenOutputStream.

**Comment by madanadit** [comment]
@aaudiber The default value for ensureAtomic is true as that is what is desired from the UFS when creating a files exisitng in the Alluxio FileSystem. The real issue is that we re-use the same API for journal and UFS files. As part of future work, I'll separate that out. 

Flush on any object store is meaningless regardless of the option. The behavior is not consistent at the moment).

File: tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java [hunk 4]
```diff
@@ -134,6 +137,7 @@ private FileSystemMaster createFsMasterFromJournal() throws IOException {
    * Reproduce the journal and check if the state is correct.
    */
   private void reproduceAndCheckState(int successFiles) throws Exception {
+    Assert.assertNotEquals(successFiles, 0);
```
Comment (calvinjia) [inline tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java]: Why is this check added?

Comment (madanadit) [inline tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java]: If successFiles are zero the listStatus below will throw an error with FileNotFounException.

File: core/common/src/main/java/alluxio/underfs/options/CreateOptions.java [hunk 5]
```diff
@@ -42,6 +46,24 @@ public Permission getPermission() {
   }
 
   /**
+   * @return ensure atomic
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/options/CreateOptions.java]: this is user-facing so we should elaborate what we mean by ensure atomic

File: core/common/src/main/java/alluxio/underfs/options/CreateOptions.java [hunk 6]
```diff
@@ -42,6 +46,24 @@ public Permission getPermission() {
...
+   */
+  public boolean isEnsureAtomic() {
+    return mEnsureAtomic;
+  }
+
+  /**
+   * Sets ensure atomic.
+   *
+   * @param atomic whether to ensure created stream is atomic
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/options/CreateOptions.java]: elaborate what it means for a stream to be atomic

File: core/common/src/main/java/alluxio/underfs/UnderFileSystem.java [hunk 7]
```diff
@@ -310,6 +310,9 @@ public OutputStream create(String path) throws IOException {
    * @throws IOException if a non-Alluxio error occurs
    */
   public OutputStream create(String path, CreateOptions options) throws IOException {
+    if (!options.isEnsureAtomic()) {
```
Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: Could we rename this to something like "writeDirect" or "direct"? The negation of "ensureAtomic" is "don't ensure anything", but the journal needs to ensure that when it flushes data that data is persisted in the file being written.

File: tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java (hunk 4 above)
Comment (calvinjia) [inline tests/src/test/java/alluxio/master/JournalShutdownIntegrationTest.java]: Is it possible for successFiles to be zero through normal operation, or is the file creator guaranteed to succeed at least once?

File: core/common/src/main/java/alluxio/underfs/UnderFileSystem.java (hunk 7 above)
Comment (madanadit) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: @aaudiber That might be difficult across UFSs as object stores do not have direct writes. !ensureAtomic actually does not ensure anything. I suggest we revisit this when the journal API is designed.

Comment (aaudiber) [inline core/common/src/main/java/alluxio/underfs/UnderFileSystem.java]: ok, we can revisit when we design the journal API

**Comment by madanadit** [comment]
@aaudiber

**Pull request merged**